src/world/GameWorldView.cpp

```diff
@@ -293,8 +293,9 @@ void GameWorldView::DrawGUI(const RoadBuildState& rb, const TerrainRenderer& ter
                 if(rb.mode == RM_BOAT && maxWaterWayLen != 0 && rb.route.size() >= maxWaterWayLen)
                     continue;
 
+                const bool targetsFlag = GetWorld().GetNO(curPt)->GetType() == NOP_FLAG && curPt != rb.start;
                 if((gwv.IsRoadAvailable(rb.mode == RM_BOAT, curPt) && gwv.IsOwner(curPt) && GetWorld().IsPlayerTerritory(curPt))
-                   || (gwv.GetBQ(curPt) == BQ_FLAG))
+                   || (gwv.GetBQ(curPt) == BQ_FLAG) || targetsFlag)
```

Comment:
I'm thinking about this `if`. The first 3 conditions ensure, that a road can be build. OK. Then we check, if we could build a flag there. But that would mean, that a road can be build, right? So we can remove that.

Problem is: `IsRoadAvailable` returns false for the point, when there is a flag (at least I think so: https://github.com/Shawn8901/s25client/blob/8e7b512a3a30d8029fb689afcbf5b933bad5f66b/src/world/GameWorldBase.cpp#L102) So that's why to `|| targetsFlag` is required. 

IMO The BQ check was a failed attempt to do that. Or am I missing a point?

Reply:
About the bug: you are 100% right.
for me the old code is a bit fishy and hard to understand, so i did a little refactoring session on that (as requested), so now the code is a bit more clear and you can follow the flow a bit easier (at least on my opinion).
After refactoring it was also easy to see why/where the bug you noticed happen (if-else-if hell :-) )

about the loop in 287: i have removed it and replaced by one of your helpers. since i am only interested if curPt is in road_points i have switched to your contains helper instead of the indexOf. Maybe there is a bit more efficient way to get the back direction, of i think my solution is quite easy to understand (i don't see any other reason to check for the index).

if there is anything else open, please let me know.